zeapp/app/src/main/java/de/berlindroid/zeapp/hardware/Badge.kt

```diff
@@ -0,0 +1,196 @@
+                            Log.e("Badge Connection", "Failed with result $result.")
+                        }
+                    } catch (th: Throwable) {
+                        Log.e("Badge Connection", "Error while sending.", th)
+                    } finally {
+                        connection.close()
+                    }
+                }
+            }
+        }
+    }
+
+    private fun askPermissionAndRetry(
+        context: Context,
+        manager: UsbManager,
+        device: UsbDevice,
+        payload: Payload
+    ) {
+        class BoundUsbReceiver : BroadcastReceiver() {
+            override fun onReceive(context: Context, intent: Intent) {
+                if (ACTION_USB_PERMISSION == intent.action) {
+                    val device: UsbDevice? = intent.getParcelableExtra(UsbManager.EXTRA_DEVICE)
+
+                    if (intent.getBooleanExtra(UsbManager.EXTRA_PERMISSION_GRANTED, false)) {
+                        device?.apply {
+                            Log.d("USB Permission", "Permission granted.")
+
+                            sendToUsb(context, payload)
+                        }
+                    } else {
+                        Log.e("USB Permission", "Could not request permission to access to badge.")
+                    }
+                }
+            }
+        }
+
+        val permissionIntent = PendingIntent.getBroadcast(
+            context,
+            ACTION_USB_PERMISSION_REQUEST_CODE,
+            Intent(ACTION_USB_PERMISSION),
+            PendingIntent.FLAG_UPDATE_CURRENT,
+        )
+
+        val filter = IntentFilter(ACTION_USB_PERMISSION)
+        context.registerReceiver(
+            BoundUsbReceiver(),
+            filter
+        )
+
+        manager.requestPermission(device, permissionIntent)
+    }
+}
+
+private fun UsbManager.findConnectedBadge() =
+    deviceList.values.firstOrNull {
+        it.productName == USB_DEVICE_PRODUCT_NAME
+    }
+
+private fun UsbManager.connectedProductNames() =
+    deviceList.values.map { it.productName ?: "<none>" }
+
+private fun UsbDevice.findParameter(): Badge.FoundUSBConnectionParameters? {
+    var foundInterface: UsbInterface? = null
+    var foundEndpoint: UsbEndpoint? = null
+
+    for (interfaceIndex in 0 until interfaceCount) {
+        val interf = getInterface(interfaceIndex)
+        for (endpointIndex in 0 until interf.endpointCount) {
+            val endpoint = interf.getEndpoint(endpointIndex)
+
+            if (endpoint.isCompatible()) {
+                foundInterface = interf
+                foundEndpoint = endpoint
+                break;
+            }
+        }
+
+        if (foundEndpoint != null) {
+            break
+        }
+    }
+
+    return if (foundInterface != null && foundEndpoint != null) {
+        Badge.FoundUSBConnectionParameters(foundInterface, foundEndpoint)
+    } else {
+        null
+    }
+}
+
+private fun UsbEndpoint.isCompatible(): Boolean = type == UsbConstants.USB_ENDPOINT_XFER_BULK
```

Comment:
@milosmns next question: How do I identify a badger endpoint? Is that xfer bulk still valid?